src/core/storage/serialization/iarchive.hpp

```diff
@@ -117,7 +117,7 @@ namespace turi {
     template <typename T>
     inline void read_into(T& c) {
       if (buf) {
-        memcpy(&c, buf + off, sizeof(T));
+        memcpy(reinterpret_cast<void*>(&c), buf + off, sizeof(T));
```

Comment:
This should not be necessary.  A cast to void* is automatic for any trivial types (i.e. no inheritance, no pointers, etc., and this memcpy is valid only for those types.  The so the lack of an explicit cast here is a guard against that (though it could be made more explicit with a 

static_assert(std::is_trivial<T>::value, "Only trivial types allowed for hash.");

Reply:
Here is the error message I get without the `reinterpret_cast`:
```
[ 14%] Building CXX object src/model_server/CMakeFiles/unity.dir/server/registration.cpp.o
In file included from /home/toby/Documents/turicreate/src/core/storage/serialization/serialize.hpp:8,
                 from /home/toby/Documents/turicreate/src/core/storage/serialization/serialization_includes.hpp:6,
                 from /home/toby/Documents/turicreate/src/core/parallel/atomic.hpp:12,
                 from /home/toby/Documents/turicreate/src/core/data/flexible_type/flexible_type.hpp:15,
                 from /home/toby/Documents/turicreate/src/model_server/lib/toolkit_class_specification.hpp:10,
                 from /home/toby/Documents/turicreate/src/model_server/lib/toolkit_class_registry.hpp:8,
                 from /home/toby/Documents/turicreate/src/model_server/server/registration.hpp:9,
                 from /home/toby/Documents/turicreate/src/model_server/server/registration.cpp:6:
/home/toby/Documents/turicreate/src/core/storage/serialization/iarchive.hpp: In instantiation of ‘void turi::iarchive::read_into(T&) [with T = boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >]’:
/home/toby/Documents/turicreate/src/core/storage/serialization/iarchive.hpp:243:9:   required from ‘static void turi::archive_detail::deserialize_impl<InArcType, T, true>::exec(InArcType&, T&) [with InArcType = turi::iarchive; T = boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >]’
/home/toby/Documents/turicreate/src/core/storage/serialization/iarchive.hpp:258:65:   required from ‘turi::iarchive& turi::operator>>(turi::iarchive&, T&) [with T = boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >]’
/home/toby/Documents/turicreate/src/core/util/hash_value.hpp:65:39:   required from here
/home/toby/Documents/turicreate/src/core/storage/serialization/iarchive.hpp:121:8: error: ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type ‘class boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >’ with no trivial copy-assignment; use copy-assignment or copy-initialization instead [-Werror=class-memaccess]
  121 |  memcpy(&c, buf + off, sizeof(T));
      |  ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/toby/Documents/turicreate/src/external/boost/boost_1_68_0/boost/multiprecision/cpp_int.hpp:12,
                 from /home/toby/Documents/turicreate/src/core/util/int128_types.hpp:44,
                 from /home/toby/Documents/turicreate/src/core/data/flexible_type/flexible_type.hpp:16,
                 from /home/toby/Documents/turicreate/src/model_server/lib/toolkit_class_specification.hpp:10,
                 from /home/toby/Documents/turicreate/src/model_server/lib/toolkit_class_registry.hpp:8,
                 from /home/toby/Documents/turicreate/src/model_server/server/registration.hpp:9,
                 from /home/toby/Documents/turicreate/src/model_server/server/registration.cpp:6:
/home/toby/Documents/turicreate/src/external/boost/boost_1_68_0/boost/multiprecision/number.hpp:41:7: note: ‘class boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void> >’ declared here
   41 | class number
      |       ^~~~~~
cc1plus: error: unrecognized command line option ‘-Wno-unknown-warning-option’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-parentheses-equality’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-constant-logical-operand’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-mismatched-tags’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-deprecated-register’ [-Werror]
cc1plus: error: unrecognized command line option ‘-Wno-unused-command-line-argument’ [-Werror]
cc1plus: all warnings being treated as errors
make[2]: *** [src/model_server/CMakeFiles/unity.dir/build.make:63: src/model_server/CMakeFiles/unity.dir/server/registration.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:3952: src/model_server/CMakeFiles/unity.dir/all] Error 2
make: *** [Makefile:152: all] Error 2
```

Reply:
The following program prints out `non-trivial`:

```c++
#include <iostream>
#include <boost/multiprecision/number.hpp>
#include <boost/multiprecision/cpp_int.hpp>

int main()
{

  const bool is_trivial = std::is_trivial<
    boost::multiprecision::number<boost::multiprecision::backends::cpp_int_backend<128, 128, boost::multiprecision::unsigned_magnitude, boost::multiprecision::unchecked, void>>
    >::value;

  if(is_trivial)
     std::cout<<"trivial"<<std::endl;
  else
    std::cout<<"non-trivial"<<std::endl;
  return 0;
}
```

Reply:
Yes, exactly, as it should.  Only trivial types can be used in that serialization function.  Boost multiprecision is the fallback type when no native 128 bit integer types are detected at build time (maybe it should error instead?).  These are detected by the Find128BitInteger.cmake routines under cmake/.  It appears these routines are failing.  That is the root cause of this problem.

Reply:
What's the output of cmake config log after a clean cmake run?

Reply:
Here is the relevant section of the logs:
```
Testing for 128 bit integer support with __int128_t.
Found: Enabling support for 128 bit integers using __int128_t.
Testing for presence of 128 bit unsigned integer hash function for __int128_t.
std::hash<__int128_t> defined.
Testing for 128 bit unsigned integer support with __uint128_t.
```
Output for debug and release is the same.

So for both signed and unsigned, it seems to be finding the first things it's looking for.

Reply:
Sorry to drop the ball on this -- let's dig into it tomorrow.  There's some typo somewhere.